File: src/nishikihebi/agents/issue_review/graph.py
from langgraph.graph import END, START, StateGraph
from langgraph.graph.state import CompiledStateGraph
from langgraph.types import RetryPolicy

from nishikihebi.agents._shared import post_review_comments
from nishikihebi.agents.issue_review.nodes import fetch_issues, review_issues
from nishikihebi.agents.issue_review.state import IssueReviewState
from nishikihebi.clients.github import GitHubClient
from nishikihebi.clients.llm import LlmClient
from nishikihebi.logs import get_logger
from nishikihebi.settings import LABEL, LABEL_COLOR, REVIEWER_LOGIN

log = get_logger(__name__)

_RETRY_POLICY = RetryPolicy(max_attempts=3)


def build_issue_review_graph(
    client: LlmClient,
    github: GitHubClient,
    reviewer_login: str = REVIEWER_LOGIN,
    label: str = LABEL,
    label_color: str = LABEL_COLOR,
) -> CompiledStateGraph:
    log.debug(
        "wiring issue_review graph nodes",
        reviewer_login=reviewer_login,
        label=label,
    )
    graph = StateGraph(IssueReviewState)
    graph.add_node(
        "fetch_issues",
        fetch_issues(github, reviewer_login, label, label_color),
        retry_policy=_RETRY_POLICY,
    )
    graph.add_node(
        "review_issues", review_issues(client), retry_policy=_RETRY_POLICY
    )
    graph.add_node(
        "post_review_comments",
        post_review_comments(github),
        retry_policy=_RETRY_POLICY,
    )
    graph.add_edge(START, "fetch_issues")
    graph.add_edge("fetch_issues", "review_issues")
    graph.add_edge("review_issues", "post_review_comments")
    graph.add_edge("post_review_comments", END)
    compiled = graph.compile()
    log.info("issue_review graph ready")
    return compiled

File: apps/server/agents/issue-review/graph.ts
import { END, START, StateGraph } from "@langchain/langgraph";
import { getLogger } from "../../../../packages/shared/logs.js";
import type { LlmClient } from "../../clients/llm.js";
import type { GitHubClient } from "../../external/github/client.js";
import { LABEL, LABEL_COLOR, REVIEWER_LOGIN } from "../../external/github/settings.js";
import { postReviewComments } from "../shared.js";
import { fetchIssues, reviewIssues } from "./nodes.js";
import { IssueReviewAnnotation } from "./state.js";

const log = getLogger("angel.agents.issue-review.graph");

const RETRY_POLICY = { maxAttempts: 3 };

export interface IssueReviewGraphOptions {
  readonly reviewerLogin?: string;
  readonly label?: string;
  readonly labelColor?: string;
}

export function buildIssueReviewGraph(
  client: LlmClient,
  github: GitHubClient,
  options: IssueReviewGraphOptions = {},
) {
  const reviewerLogin = options.reviewerLogin ?? REVIEWER_LOGIN;
  const label = options.label ?? LABEL;
  const labelColor = options.labelColor ?? LABEL_COLOR;

  log.debug("wiring issue_review graph nodes", {
    reviewer_login: reviewerLogin,
    label,
  });
  const graph = new StateGraph(IssueReviewAnnotation)
    .addNode("fetch_issues", fetchIssues(github, reviewerLogin, label, labelColor), {
      retryPolicy: RETRY_POLICY,
    })
    .addNode("review_issues", reviewIssues(client), { retryPolicy: RETRY_POLICY })
    .addNode("post_review_comments", postReviewComments(github), {
      retryPolicy: RETRY_POLICY,
    })
    .addEdge(START, "fetch_issues")
    .addEdge("fetch_issues", "review_issues")
    .addEdge("review_issues", "post_review_comments")
    .addEdge("post_review_comments", END);

  const compiled = graph.compile();
  log.info("issue_review graph ready");
  return compiled;
}
